libs/wingc/src/type_check.rs

```diff
@@ -216,177 +267,79 @@ impl Display for Type {
-				write!(f, "instance of {}", class_type.name.name)
-			}
-			Type::Namespace(namespace) => write!(f, "{}", namespace.name),
 			Type::Struct(s) => write!(f, "{}", s.name),
-			Type::StructInstance(s) => {
-				let struct_type = s.as_struct().expect("Struct instance must reference to a struct");
-				write!(f, "instance of {}", struct_type.name.name)
-			}
 			Type::Array(v) => write!(f, "Array<{}>", v),
 			Type::Map(v) => write!(f, "Map<{}>", v),
 			Type::Set(v) => write!(f, "Set<{}>", v),
 			Type::Enum(s) => write!(f, "{}", s.name),
-			Type::EnumInstance(e) => {
-				write!(f, "enum value {}.{}", e.enum_name, e.enum_value)
-			}
 		}
 	}
 }
 
-//type TypeRef = *const Type;
-
-#[derive(Clone, Copy)]
-pub struct TypeRef(*const Type);
-
 // TODO Allows for use in async runtime
 // TODO either avoid shared memory or use Arc<Mutex<...>> instead
 unsafe impl Send for TypeRef {}
 
-impl From<&Box<Type>> for TypeRef {
-	fn from(t: &Box<Type>) -> Self {
-		Self(&**t as *const Type)
-	}
-}
-
-impl From<TypeRef> for &Type {
-	fn from(t: TypeRef) -> Self {
-		unsafe { &*t.0 }
-	}
-}
-
-impl From<TypeRef> for &mut Type {
-	fn from(t: TypeRef) -> Self {
-		unsafe { &mut *(t.0 as *mut Type) }
-	}
-}
-
 impl TypeRef {
-	pub fn as_class_or_resource_object(&self) -> Option<&Class> {
-		self.as_class_object().or_else(|| self.as_resource_object())
-	}
-
-	pub fn as_class_object(&self) -> Option<&Class> {
-		if let &Type::ClassInstance(ref class_instance) = (*self).into() {
-			let class = class_instance.as_class().unwrap();
-			Some(class)
-		} else {
-			None
-		}
-	}
-
-	pub fn as_resource_object(&self) -> Option<&Class> {
-		if let &Type::ResourceObject(ref res_obj) = (*self).into() {
-			let res = res_obj.as_resource().unwrap();
+	pub fn as_resource(&self) -> Option<&Class> {
+		if let Type::Resource(ref res) = **self {
 			Some(res)
 		} else {
 			None
 		}
 	}
 
-	pub fn as_resource(&self) -> Option<&Class> {
-		if let &Type::Resource(ref res) = (*self).into() {
-			Some(res)
-		} else {
-			None
-		}
+	pub fn as_class_or_resource(&self) -> Option<&Class> {
+		self.as_class().or_else(|| self.as_resource())
 	}
 
 	fn as_class(&self) -> Option<&Class> {
-		if let &Type::Class(ref class) = (*self).into() {
+		if let Type::Class(ref class) = **self {
 			Some(class)
 		} else {
 			None
 		}
 	}
 
 	fn as_struct(&self) -> Option<&Struct> {
-		if let &Type::Struct(ref s) = (*self).into() {
+		if let Type::Struct(ref s) = **self {
 			Some(s)
 		} else {
 			None
 		}
 	}
 
-	fn as_enum(&self) -> Option<&Enum> {
-		if let &Type::Enum(ref e) = (*self).into() {
-			Some(e)
-		} else {
-			None
-		}
-	}
-
-	fn as_mut_enum(&self) -> Option<&mut Enum> {
-		if let &mut Type::Enum(ref mut e) = (*self).into() {
-			Some(e)
-		} else {
-			None
-		}
-	}
-
 	fn maybe_unwrap_option(&self) -> TypeRef {
-		if let &Type::Optional(ref t) = (*self).into() {
-			*t
+		if let Type::Optional(ref t) = **self {
+			t.clone()
```

Comment:
Why clone here? Wouldn't this work still?
```suggestion
			*t
```